backend/eventloop/cluster_reconciler.go

```diff
@@ -0,0 +1,206 @@
+//+kubebuilder:rbac:groups="apps",resources=replicasets;statefulsets;daemonsets;deployments,verbs=get;list;delete
+//+kubebuilder:rbac:groups="discovery.k8s.io",resources=endpointslices,verbs=get;list;delete
+//+kubebuilder:rbac:groups="networking.k8s.io",resources=ingresses;ingressclasses,verbs=get;list;delete
+//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings;roles,verbs=get;list;delete
+//+kubebuilder:rbac:groups="route.openshift.io",resources=routes,verbs=get;list;delete
+//+kubebuilder:rbac:groups="triggers.tekton.dev",resources=eventlisteners;triggertemplates,verbs=get;list;delete
+//+kubebuilder:rbac:groups="pipelinesascode.tekton.dev",resources=repositories,verbs=get;list;delete
+
+func (c *ClusterReconciler) Start() {
+	go func() {
+		<-time.NewTimer(orphanedResourcesCleanUpInterval).C
+
+		ctx := context.Background()
+		log := log.FromContext(ctx).
+			WithName(logutil.LogLogger_managed_gitops).
+			WithValues("component", "cluster-reconciler").
+			WithValues(sharedutil.JobKey, sharedutil.JobKeyValue)
+
+		_, _ = sharedutil.CatchPanic(func() error {
+			c.cleanOrphanedResources(ctx, log)
+			return nil
+		})
+
+		// Kick off the timer again, once the old task runs.
+		c.Start()
+	}()
+}
+
+// A k8s resource is considered to be orphaned when:
+// 1. It was previously managed by Argo CD i.e has label "app.kubernetes.io/instance".
+// 2. It doesn't have a corresponding GitOpsDeployment resource.
+func (c *ClusterReconciler) cleanOrphanedResources(ctx context.Context, log logr.Logger) {
+	// Use a label selector to filter resources managed by Argo CD
+	labelSelector, err := labels.Parse(argocdResourceLabel)
+	if err != nil {
+		log.Error(err, "failed to create a label selector for listing resources managed by Argo CD")
+		return
+	}
+
+	listOpts := &client.ListOptions{
+		LabelSelector: labelSelector,
+	}
+
+	apiObjects, err := c.getAllNamespacedAPIResources(ctx, log, listOpts)
+	if err != nil {
+		log.Error(err, "failed to get namespaced API resources from the cluster")
+		return
+	}
+
+	for i, obj := range apiObjects {
+		appName := getArgoCDApplicationName(obj.GetLabels())
+		if appName != "" && obj.GetDeletionTimestamp() == nil && !strings.HasPrefix(obj.GetNamespace(), "openshift") {
+			// Check if a GitOpsDeployment exists with the UID specified in the Application name.
+			gitopsDeplList := &managedgitopsv1alpha1.GitOpsDeploymentList{}
+			err = c.client.List(ctx, gitopsDeplList, &client.ListOptions{
+				Namespace: obj.GetNamespace(),
+			})
+			if err != nil {
+				log.Error(err, "failed to list GitOpsDeployments", "namespace", obj.GetNamespace())
+				continue
+			}
+
+			expectedUID := extractUIDFromApplicationName(appName)
+			found := false
+			for _, gitopsDepl := range gitopsDeplList.Items {
+				if gitopsDepl.UID == types.UID(expectedUID) {
+					found = true
+					break
```

Comment:
This block appears not to be covered by the unit tests (correct me if I'm wrong!). It couldn't hurt to make sure that a resources is NOT deleted when there is a valid GitOpsDeployment existing for it: there IS a test for this scenario in the PR ("should not delete a resource that is still managed by GitOpsDeployment"), but it would be good to double check why that test isn't covering this block.

Reply:
The test `should not delete a resource that is still managed by GitOpsDeploymen` will cover this case. It has a GitOpsDeployment that matches the ID from the Application so it is not deleted and the test checks that the resource still exists.

Reply:
This is what I'm seeing in 'cleanOrphanedResources', after running the tests and applying the cover.out profile in VS Code:

![image](https://github.com/redhat-appstudio/managed-gitops/assets/7594762/2835b6ee-cdcf-4abd-b3b5-c5ff4fb85632)

Based on this, it seems like `gitopsDeplList` never has a value (and thus also the UID check isn't tested).

Reply:
Fixed the issue. The test was not creating the resource properly. The code block is now covered 

![Screenshot from 2023-08-10 11-42-29](https://github.com/redhat-appstudio/managed-gitops/assets/21128732/be712ab1-d42d-4311-877f-a1604eea0b20)